Add List<T>.CopyTo Span<T> APIs

felipepessoto opened the pull request:
Add List<T>.CopyTo(Span<T> destination) and List<T>.CopyTo(int sourceIndex, int count, Span<T> destination)

Fix https://github.com/dotnet/corefx/issues/33006

-----
I watched the Design Review video but I still have some questions (this is my first non-trivial contribution).

1. CopyTo(Span<T> destination) should throw if Span.Length is smaller than List.Count?
2. CopyTo(int sourceIndex, int count, Span<T> destination) should throw if count is greater than destination Span?
3. Do I need to use the ThrowHelper.ThrowXXX methods? Or can I just use throw new XException?
4. Is expected to use Buffer.Memmove? I'm not familiar with this class.

Thanks

@safern @sywhang

felipepessoto committed:
Add List<T>.CopyTo Span<T> APIs

Add List<T>.CopyTo(Span<T> destination) and List<T>.CopyTo(int sourceIndex, int count, Span<T> destination)

Fix https://github.com/dotnet/corefx/issues/33006

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +402,39 @@ public void CopyTo(T[] array, int arrayIndex)
 
+        public void CopyTo(int sourceIndex, int count, Span<T> destination)
+        {
+            if (sourceIndex < 0)
```
Given this extensive argument validation, this API is likely going to slower (for small number of elements) than `CollectionMarshal.AsSpan(list).CopyTo(span)` that you can do today already. It makes me wonder whether these APIs are really worth it - added comment at https://github.com/dotnet/corefx/issues/33006#issuecomment-565879150.

felipepessoto replied:
Where can I find the CollectionMarshal?

felipepessoto replied:
You would need an additional Slide() what makes more validation and CopyTo also make a validation.
We are probably making less validations here, but I don't know if I would add this API (even if it is faster and more convenient) if we already have CollectionMarshal.AsSpan.

jkotas replied:
> Slice() what makes more validation

Note that the validation in the Span methods is fine-tuned using various tricks. For example, `Slice` avoids one of the checks via a trick that requires 6 line comment to explain. (I do not think we want to be duplicating these tricks here or in other places.)

felipepessoto commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +402,39 @@ public void CopyTo(T[] array, int arrayIndex)
+
+            if (_size > 0 && count > 0)
+            {
+                Buffer.Memmove(ref destination[0], ref _items[sourceIndex], (nuint)count);
```
Should I change it to:

Buffer.Memmove(ref destination._pointer.Value, ref Unsafe.Add(ref Unsafe.As<byte, T>(ref _items.GetRawSzArrayData()), sourceIndex), (nuint)count);

?

felipepessoto commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +402,39 @@ public void CopyTo(T[] array, int arrayIndex)
+            }
+        }
+
+       public void CopyTo(Span<T> destination)
```
If we decide to add this APIs, I can implement this overload avoiding most of the argument validation.

felipepessoto commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +402,39 @@ public void CopyTo(T[] array, int arrayIndex)
+                throw new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination));
+            }
+
+            if (_size > 0 && count > 0)
```
I guess I don't to check size > 0. If count is 1, size should be 1+, or else it would throw in if (_size - sourceIndex < count)

jkotas replied:
The standard we code most libraries to is that race condition in the app should not lead to buffer overrun. It means that you cannot depend on `_size <= _items.Length`. They can get out of sync if there is a race condition.

It would be best to just call `_items.AsSpan(sourceIndex, count).CopyTo(destination)` here. It will be easy to see that there is no buffer overrun that way.

felipepessoto committed:
Change the implementation to use AsSpan

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +393,16 @@ public void CopyTo(T[] array, int arrayIndex)
         }
 
+        public void CopyTo(int sourceIndex, int count, Span<T> destination)
+        {
```
The argument validation would be still needed here to prevent copying more than `List.Length` items and to use the right argument names in exception messages.

felipepessoto replied:
Done

felipepessoto committed:
Argument validations

Fix CopyTo(Span<T>) overload to pass the right size

Add test to ensure Copy Span doesn't copy more than List.Count when the backing array is larger than List

felipepessoto committed:
Matches argument names with implementation

felipepessoto commented:
@jkotas is it my PR, or CI is broken?

jkotas commented:
CI is broken.

felipepessoto commented:
/azp run

felipepessoto commented:
/azp run runtime-coreclr

safern commented:
/azp run runtime-coreclr

safern commented:
/azp run runtime-libraries

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -393,6 +393,31 @@ public void CopyTo(T[] array, int arrayIndex)
             Array.Copy(_items, 0, array, arrayIndex, _size);
         }
 
+        public void CopyTo(int sourceIndex, int count, Span<T> destination)
```
@felipepessoto Your PR is adding new public APIs in the ref file. Please make sure to document them with triple slash comments so the reviewers can sign off your change. We want to make sure we always merge new public APIs with documentation.

felipepessoto replied:
@carlossanlop, we are still adding this API?

carlossanlop replied:
Thanks for documenting the public APIs with triple slash comments.
I'm not sure I understand the question. The purpose of this PR is to add these two APIs, so yes?

felipepessoto committed:
Comments

carlossanlop commented:
The new public APIs have been properly documented with triple slash comments. Thanks @felipepessoto. Removing the label.
Adding @safern as reviewer for System.Collections.

safern commented:
> @carlossanlop, are we still adding this API?

I spoke with @eerhardt and @jkotas and I think we agreed that it makes sense to not add this APIs given we can achieve the same using CollectionsMarshal and also that in order to have a "complete" span support and also, adding these APIs might slow down other operations on `List<T>` so it seems reasonable to teach people to use `CollectionsMarshal` instead. 

Sorry @felipepessoto for this and thanks for the effort on putting this together, looking forward to your next contribution 😄 -- will close this, if people disagree, we can of course re-open.
